services/harmony/app/backends/workflow-orchestration/work-item-updates.ts

```diff
@@ -445,68 +446,80 @@ async function createNextWorkItems(
+  //  a new batch if neeeded. We have a check to see if we have completed all the work-items for
+  //  the current step, in which case we close the last batch even if it is not full and
+  //  generate a final work-item for the next step
+  // 2. next step aggregates, but does not use batching, so we only create a new work-item
+  //  for the next step if we have completed all the work-items for the current step
+  // 3. next step does not aggregate, so we create a new work-item for the next step for each
+  //  result from this work-item update
+  // if (results && results.length > 0 || nextWorkflowStep.isBatched) {
+  if (nextWorkflowStep.hasAggregatedOutput) {
+    if (nextWorkflowStep.isBatched) {
+      let sortIndex;
+      if (!QUERY_CMR_SERVICE_REGEX.test(workItem.serviceID)) {
+        // eslint-disable-next-line prefer-destructuring
+        sortIndex = workItem.sortIndex;
       }
-    } else {
-      // Create a new work item for each result using the next step
-
-      // use the sort index from the previous step's work item unless the service was
-      // query-cmr, in which case we start from the previous highest sort index for this step
-      // NOTE: This is only valid if the work-items for this multi-output step are worked
-      // sequentially and have consistently ordered outputs, as with query-cmr.
-      // If they are worked in parallel then we need a different approach.
-      let { sortIndex } = workItem;
-      let shouldIncrementSortIndex = false;
-      if (QUERY_CMR_SERVICE_REGEX.test(workItem.serviceID)) {
-        shouldIncrementSortIndex = true;
-        sortIndex = await maxSortIndexForJobService(tx, nextWorkflowStep.jobID, nextWorkflowStep.serviceID);
+      let outputItemUrls = [];
+      if (workItem.status !== WorkItemStatus.FAILED) {
+        outputItemUrls = await outputStacItemUrls(results);
       }
-      const newItems = results.map(result => {
-        if (shouldIncrementSortIndex) {
-          sortIndex += 1;
-        }
-        return new WorkItem({
-          jobID: workItem.jobID,
-          serviceID: nextWorkflowStep.serviceID,
-          status: WorkItemStatus.READY,
-          stacCatalogLocation: result,
-          workflowStepIndex: nextWorkflowStep.stepIndex,
-          sortIndex,
-        });
+      didCreateWorkItem = await handleBatching(
+        tx,
+        logger,
+        nextWorkflowStep,
+        outputItemUrls,
+        outputItemSizes,
+        sortIndex,
+        workItem.status,
+        allWorkItemsForStepComplete);
+    } else if (allWorkItemsForStepComplete) {
+      await createAggregatingWorkItem(tx, workItem, nextWorkflowStep, logger);
+      didCreateWorkItem = true;
+    }
+  } else {
+    // Create a new work item for each result using the next step
+    didCreateWorkItem = true;
+
+    // use the sort index from the previous step's work item unless the service was
+    // query-cmr, in which case we start from the previous highest sort index for this step
```

Comment:
Is this comment still correct that it only applies to query-cmr? I guess most importantly does this logic still work for batchee, stitchee, and concise?

Reply:
query-cmr is the only sequential service we have right now, and we only use that logic for sequential services. To answwer your other question, yes, this works with batchee, stitchee, and concise.